vdom/src/com/vdom/core/PileCreator.java

```diff
@@ -0,0 +1,97 @@
+package com.vdom.core;
+
+import com.vdom.api.Card;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Random;
+
+public abstract class PileCreator implements Serializable {
+    public abstract AbstractCardPile create(Card template, int count);
+}
+
+class DefaultCardPileCreator extends PileCreator {
+    public AbstractCardPile create(Card template, int count) {
+        return new SingleCardPile(template, count);
+    }
+}
+
+class RuinsPileCreator extends PileCreator {
+    public AbstractCardPile create(Card template, int count) {
+
+        Map<Card, Integer> cards = new HashMap<Card, Integer>();
+
+        Random rand = new Random();
+        ArrayList<Integer> counts = new ArrayList<Integer>();
+
+        //Generate n-1 numbers between 0 and count
+        for (int i = 0; i < Cards.ruinsCards.size()-1; i++) {
+            counts.add(rand.nextInt(count));
+        }
+        // Add 0 and count. Then sort the list.
+        counts.add(0);
+        counts.add(count);
+        Collections.sort(counts);
+        //Then sort the list and use the adjacent numbers difference as counts for the particular ruins.
+        for (int i = 0; i < Cards.ruinsCards.size(); i++) {
+            cards.put(Cards.ruinsCards.get(i), counts.get(i+1) - counts.get(i));
+        }
+
+        return new VariableCardPile(template, cards, false, false);
+    }
+}
+
+class KnightsPileCreator extends PileCreator {
+
+    public AbstractCardPile create(Card template, int count) {
+        Map<Card, Integer> cards = new HashMap<Card, Integer>();
+        //Currently count is ignored because there should always be ten knights.
+        for (Card c: Cards.knightsCards) {
+            cards.put(c, 1);
+        }
+        return new VariableCardPile(template, cards, false, false);
+
+    }
+}
+
+class CastlesPileCreator extends PileCreator {
+    public AbstractCardPile create(Card template, int count) {
+        Map<Card, Integer> cards = new LinkedHashMap<Card, Integer>(); //LinkedHashMap preserves insertion order when iterating
+        if (count != 8 && count != 12) {
+            //TODO SPLITPILES What to do now?
+            if (count < 8) count = 8;
+            if (count > 8) count = 12;
+        }
+        int multiCastles = (count == 8 ? 1 : 2);
+        boolean multi = true;
+        for (Card c: Cards.castleCards) {
+            cards.put(c, multi ? multiCastles : 1); //if there are 12 cards, add two for every second card (start at first)
```

Comment:
This is wrong - there are 2 King's castles, not two Grand Castles. I would recommend just listing out the kinds of Castles that have 2 copies (Humble, Small, Opulent and King's) instead and using that.